lib/get/walkPath.js

```diff
@@ -16,58 +16,40 @@ module.exports = function walkPath(model, root, curr, path, depth, seed,
     var optimizedPath = optimizedPathArg;
     var referenceContainer = referenceContainerArg;
 
-    // If there is not a value in the current cache position or its a
-    // value type, then we are at the end of the getWalk.
-    if ((!curr || curr && curr.$type) || depth === path.length) {
+    var keySet = path[depth];
+    var isKeySet = keySet !== null && typeof keySet === "object";
+    var iteratorNote = isKeySet ? {} : undefined;
+    var key = isKeySet ? iterateKeySet(keySet, iteratorNote) : keySet;
+
+    // The walk is finished when:
+    // - there is no value in the current cache position
+    // - there is a JSONG leaf node in the current cache position
+    // - we've reached the end of the path
+    // - the current key set contains no keys
+    if (!curr || curr.$type || depth === path.length || (isKeySet && iteratorNote.done)) {
         onValueType(model, curr, path, depth, seed, outerResults, branchInfo,
                 requestedPath, optimizedPath, optimizedLength,
                 isJSONG, fromReference);
         return;
     }
 
-    var i;
-    var keySet = path[depth];
-    var isKeySet = typeof keySet === "object";
-    var nextDepth = depth + 1;
-    var iteratorNote = false;
-    var key = keySet;
     var allowFromWhenceYouCame = model._allowFromWhenceYouCame;
-
-    if (isKeySet) {
-        iteratorNote = {};
-        key = iterateKeySet(keySet, iteratorNote);
-    }
-
-    // loop over every key over the keySet
     var optimizedLengthPlus1 = optimizedLength + 1;
+    var nextDepth = depth + 1;
     var refPath;
-    do {
-        fromReference = false;
-
-        var next;
-
-        // There are two cases when it comes to a null key.  In path vs at the
-        // end of a path.
-        if (key === null) {
-            // If the key is null and we are not at the end of a path, then
-            // throw an error.
-            if (depth < path.length) {
-                throw new NullInPathError();
-            }
+    var i;
 
-            // Else, we are at the end of a path, then just say next is current.
-            else {
-                next = curr;
-            }
+    // loop over every key in the key set
+    do {
+        if (key == null) {
+            throw new NullInPathError();
         }
 
-        // The standard case, do the depth search into the cache.
-        else {
-            next = curr[key];
-            optimizedPath[optimizedLength] = key;
-            requestedPath[depth] = key;
-        }
+        fromReference = false;
+        optimizedPath[optimizedLength] = key;
+        requestedPath[depth] = key;
```

Comment:
This is where `key` would be `undefined` which eventually ends up in missing paths.